deployment/scripts/upgrade.sh

```diff
@@ -0,0 +1,410 @@
+  for dc in "${kibana_dc[@]}"; do
+    oc scale --replicas=1 dc $dc
+
+    pod=$(getPodName $dc)
+    waitForStart $pod
+    checkKibanaStarted $pod
+  done
+
+  # TODO: if the daemonset doesn't exist, recreate it here
+  #  if the template doesn't exist, recreate it first?
+}
+
+function updateImages() {
+
+  # create any missing imagestreams and then update them all
+  oc new-app logging-imagestream-template || : # these may fail if created independently; that's ok
+
+  # TODO: update these to fail if we can't import?
+  oc import-image logging-elasticsearch || :
+  oc import-image logging-fluentd || :
+  oc import-image logging-kibana || :
+  oc import-image logging-curator || :
+}
+
+# we can just specify to regenerate our certs
+function add_curator() {
+
+  echo "Adding curator..."
+  # this may fail otherwise if there was only a partial upgrade previously
+  oc delete dc,rc,template --selector logging-infra=curator || :
+
+  generate_curator_template
+  generate_curator
+}
+
+function fluentd_daemonset() {
+  echo "Updating Fluentd to use Daemonset..."
+
+  # this may fail otherwise if there was only a partial upgrade previously
+  oc delete dc,rc,template,daemonset --selector logging-infra=fluentd || :
+
+  generate_fluentd_template
+  generate_fluentd
+}
+
+function upgrade_notify() {
+  set +x
+  echo "Upgrade complete!"
+  cat <<EOF
+
+=================================
+
+The deployer has created additional secrets, templates, and component deployments
+as part of the upgrade process. You may now have a few more steps to run manually.
+Consult the deployer docs for more detail.
+
+Fluentd:
+--------------
+Fluentd is deployed to nodes via a DaemonSet. If this is new for your deployment:
+as a cluster admin label the nodes you'd like to deploy it to:
+    oc label node/<node-name> ${fluentd_nodeselector}
+
+To label all nodes at once:
+    oc label nodes --all ${fluentd_nodeselector}
+
+Note: if your previous deployment used a DaemonSet for Fluentd, there should be
+no additional actions to deploy your pods -- the deployer did not unlabel any nodes.
+EOF
+}
+
+function regenerate_secrets_and_support_objects() {
+  oc process logging-support-template | oc delete -f - || :
+  oc delete imagestream,service,route,template --selector logging-infra=support
+  # note: dev builds aren't labeled and won't be deleted. if you need to preserve imagestreams, you can just remove the label.
+  # note: no automatic deletion of persistentvolumeclaim; didn't seem wise
+
+  generate_secrets
+  generate_support_objects
+}
+
+function upgrade_logging() {
+
+  installedVersion=$(getDeploymentVersion)
+  local migrate=
+
+  # VERSIONS
+  # 0 -- just EFK
+  # 1 -- admin cert
+  # 2 -- curator & daemonset
+
+  # start with installed version and go until we're at $LOGGING_VERSION
+  if [[ $installedVersion -eq $LOGGING_VERSION ]]; then
+    echo "Already at latest version. No upgrade will be done at this time."
```

Comment:
Here's a thought; say they're at 3.2.1 and what they really want is to just move to 3.2.2 (no structural changes, let's hope). Could "upgrade" mode do that for them? They've specified IMAGE_VERSION=3.2.2 so scale everything down, go into the DCs and patch them (or just import updated image stream tags, if the version hasn't changed) and recreate the daemonset...? Otherwise they'll have to do all that manually.